rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
@th3hamm0r no worries. Your interruption ensure that we don't make silly mistakes :).

Sorry for not being clear, perhaps I should have made it clear in my previous suggestion that we wouldn't call `pipeline.execute()` on the book keeping stuff so we can enqueue dependents and commit changes to job in one single transaction. I was thinking something like:
```python
self.set_current_job_id(None, pipeline=pipeline)

if result_ttl != 0:
    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
    job.save(pipeline=pipeline)

finished_job_registry = FinishedJobRegistry(job.origin, self.connection)
job.cleanup(result_ttl, pipeline=pipeline, remove_from_queue=False)
started_job_registry.remove(job, pipeline=pipeline)

while True:
    try:
        queue.enqueue_dependents(job, pipeline=pipeline)
        pipeline.execute()
    except WatchError:
        continue
```
But I suppose it wouldn't work either because as you pointed out, we have to replay all statements when `WatchError` occurs.

Would this work then?
```python
while True:

    self.set_current_job_id(None, pipeline=pipeline)

    if result_ttl != 0:
        job.set_status(JobStatus.FINISHED, pipeline=pipeline)
        job.save(pipeline=pipeline)

    finished_job_registry = FinishedJobRegistry(job.origin, self.connection)
    job.cleanup(result_ttl, pipeline=pipeline, remove_from_queue=False)
    started_job_registry.remove(job, pipeline=pipeline)

    try:
        pipeline.watch(job.dependents_key)
        queue.enqueue_dependents(job, pipeline=pipeline)
        pipeline.execute()
    except WatchError:
        continue

    break
```
The scope of `try/except` is much smaller since `enqueue_dependents` is the only place where the exception could be raised. This also allows dependents to be added until right before the dependents are actually enqueued.